docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+
+### Update your pull request
+
+As you get make progress on your feature or bugfix, your pull request, once submitted, will be updated each time you [push commits][self-push-commits] to your remote branch. This means you can keep your pull request open as long as you need, rather than closing and opening new ones for the same feature or bugfix.
+
+It's a good idea to keep your pull request mergeable with Zulip upstream by frequently fetching, rebasing, and pushing changes. See [keep your fork up to date][self-keep-up-to-date] for details. You might also find this excellent article [How to Rebase a Pull Request][edx-howto-rebase-pr] helpful.
+
+## Collaborate
+
+### Fetch another contributor's branch
+
+What happens when you would like to collaborate with another contributor and they have work-in-progress on their own fork of Zulip? No problem! Just add their fork as a remote and pull their changes.
+
+```
+$ git remote add <username> https://github.com/<username>/zulip.git
+$ git fetch <username>
+```
+
+Now you can checkout their branch just like you would any other. You can name the branch anything you want, but using both the username and branch name will help you keep things organized.
+
+```
+$ git checkout -b <username>-<branchname>
+```
+
+### Checkout a pull request locally
+
+Just as you can checkout any user's branch locally, you can also checkout any pull request locally. GitHub provides a special syntax ([details][github-help-co-pr-locally]) for this since pull requests are specific to GitHub rather than Git.
+
+First, fetch and create a branch for the pull request, replacing *ID* and *BRANCHNAME* with the ID of the pull request and your desired branch name:
+
+```
+$ git fetch upstream pull/ID/head:BRANCHNAME
+```
+
+Now switch to the branch:
+
+```
+$ git checkout BRANCHNAME
+```
+
+Now you work on this branch as you would any other.
+
+## Review changes
+
+### Changes on (local) working tree
+
+Display changes between index and working tree (what is not yet staged for commit):
+
+```
+$ git diff
+```
+
+Display changes between index and last commit (what you have staged for commit):
+
+```
+$ git diff --cached
+```
+
+Display changes in working tree since last commit (changes that are staged as well as ones that are not):
+
+```
+$ git diff HEAD
+```
+
+### Changes within branches
+
+Use any git-ref to compare changes between two commits on the current branch.
+
+Display changes between commit before last and last commit:
+
+```
+$ git diff HEAD^ HEAD
+```
+
+Display changes between two commits using their hashes:
+
+```
+$ git diff e2f404c 7977169
+```
+
+### Changes between branches
+
+Display changes between tip of topic branch and tip of master branch:
+
+```
+$ git diff topic master
+```
+
+Display changes that have occured on master branch since topic branch was created:
+
+```
+$ git diff topic...master
```

Comment:
I'd somewhere in here give the example of `git diff upstream/master..HEAD` as a good way to see what you've changed on a branch so far.